Initial multi-gateway support

sskaur opened the pull request:
This commit allows for coordination of the target configuration
within all gateways in a gateway group. Gateways register a watch
on the group's OMAP object and send a notify signal to the object
when they make changes to the configuration. Gateways check for
updates to the OMAP upon receiving the watch callback as well as
periodically according to an update_interval config parameter.

Signed-off-by: Sandy Kaur <sandy.kaur@ibm.com>

PepperJo commented on Makefile:
```diff
@@ -0,0 +1,31 @@
+CONFFILE ?= main.config
+
+setup: requirements.txt
+	pip install -r requirements.txt
```
Use `pip3` in case someone still uses python2 as default

PepperJo commented on control/cli.py:
```diff
@@ -99,23 +99,23 @@ def logger(self):
         return self._logger
 
-    def connect(self, nvme_config):
+    def connect(self, settings):
```
Nit: I would try to be consistent with the name, i.e. config vs settings. It was not immediately clear to me if settings refers to the content of the config file or something else.

sskaur replied:
I see some places I can more consistently call it "settings" (specifically in __main__.py & settings.py). My goal was to separate the naming in reference to the user settings in the gateway "config" file from the persistent "config". I don't know that I could be more consistent in cli.py without ceasing to refer to the config file as a "config" file. Do you mean you think it should no longer be called "main.config" and we should refer to it as a "settings" file?

PepperJo replied:
Yes either we call everything related to the local config file settings or config but I think the mix is confusing. I think changing it to `settings` everywhere makes sense to avoid confusion with the OMAP/in-memory config.

PepperJo commented on control/db.py:
```diff
@@ -118,17 +118,29 @@ def __init__(self, nvme_config):
             self.logger.error(f"Unable to write to omap: {ex}. Exiting!")
             raise
 
+        # Watch OMAP object for changes
```
Move into its own function to allow to also run without watch/notify, e.g. just polling

sskaur replied:
Sure, determined by a config file parameter? `watch_notify = False`?

PepperJo replied:
Maybe `omap_watch_notify` or `config_watch_notify` or should we be more descriptive, i.e. like `config_update_notify` ?

sskaur replied:
I like `config_update_notify` especially as we'll have `config_update_interval_sec`

PepperJo commented on control/db.py:
```diff
@@ -94,16 +94,16 @@ class OmapPersistentConfig(PersistentConfig):
     LISTENER_PREFIX = "listener_"
 
-    def __init__(self, nvme_config):
+    def __init__(self, settings):
```
It looks like this class includes 3 different concepts that IMO should be separated: local config, omap config, update
My suggestion is to change the name of the base class `PersistentConfig` to just `Config` and have two sub-classes: `LocalConfig` and `OmapConfig`. Add a function to retrieve the current version to the base. Add another sub-class to `Config` that e.g. is called `ConsistentConfig` where the constructor takes two configs as arguments where in this example one would be local the other omap. All calls like e.g. add_bdev would funnel through the `ConsistentConfig` class which updates both underlying configs and makes sure they are aligned, i.e. have the same version. The whole update code can also be put in this `ConsistentConfig` class. To compute the diff you can add another function that allows reading all keys from either.

PepperJo commented on control/db.py:
```diff
@@ -301,17 +347,51 @@ def delete_config(self):
     def restore(self, callbacks):
         """Restores gateway config to persistent config specifications."""
 
+        self.gateway_rpcs = callbacks
```
Do we still need the restore function? This can all be handled by `update` or?

PepperJo commented on control/db.py:
```diff
@@ -301,17 +347,51 @@ def delete_config(self):
+            # Handle OMAP removals
+            removed_keys = self.config.keys() - omap.keys()
+            removed = {key: self.config[key] for key in removed_keys}
+            self._update_caller(removed, False)
```
I see potential issues with the order. For example, if a object is going to be removed which is currently still referenced by one of the other objects that only changed, it will fail to remove.

PepperJo commented on control/server.py:
```diff
@@ -0,0 +1,279 @@
+            if not alive:
+                break
+
+    def _restore_config(self):
```
See above. I don't think we need restore anymore. We can just do an update.

PepperJo commented on control/server.py:
```diff
@@ -0,0 +1,279 @@
+
+    def _update_config(self):
+        """Periodically check for persistent config updates."""
+        update_interval = self.settings.getfloat("gateway", "refresh")
```
Let's not use a float value in the user provided config to specify an interval. We need a defined time unit and rename the config parameter accordingly, e.g. `config_update_interval_sec` and have a way to turn it of e.g. by setting it to 0 or -1.

PepperJo commented on control/settings.py:
```diff
@@ -0,0 +1,55 @@
+import logging
+
+
+class Settings:
```
I'm contemplating if we should use the same model as we do for the `PersistentConfig` and instead of having generic `get` functions have specific functions like `get_gateway_address` etc. This would keep all the config options in one file and would allow to sanitize the config according to the type. At least we should use constants defined in here to access the `Settings`. (As stated above let's make sure we use config vs settings consistently)

sskaur replied:
I like how it would allow us to set defaults in one place. I'm for it. Can you expand on "At least we should use constants defined in here to access the `Settings`"? I'm not sure what you mean.

PepperJo replied:
I meant we should use constants for the config keys in the Settings class, e.g. GATEWAY_ADDRESS_KEY = gateway_addr etc.

sskaur replied:
Moving further discussion of this to #49

PepperJo reviewed (changes requested):
Thank you for the PR Sandy.
Just generally, in the future I think you can make the life of the reviewers easier if you split the various changes in multiple PRs, e.g. have one for the module change, one for the Settings change etc. It also makes live for you easier eventually since now you have changes for various different things in this PR.

My main request for this PR is trying to untangle the `OmapPersistentConfig`.

PepperJo commented on control/server.py:
```diff
@@ -243,3 +265,12 @@ def _ping(self):
+
+    def _update_state(self):
+        """Periodically check for gateway state updates."""
+        update_interval = self.config.getint("gateway", "state_update_interval_sec")
```
We should define a range of valid values, e.g. what happens if someone specifies a negative value? Also should we allow to set `0` to indicate no periodic update? (Put a big warning log message in this case).

PepperJo replied:
We discussed this today in the community call and decided to not allow to disable the polling since it is always required to keep the state consistent.

PepperJo commented on control/server.py:
```diff
@@ -127,10 +133,37 @@ def serve(self):
         self.logger.info("Starting server...")
         self.server.start()
+        # Start checking for persistent configuration updates
+        self.update_timer = threading.Thread(target=self._update_state)
```
Do we want to move the `update_timer` to the consistent gateway state? Since we also update state with notify/update as part of the OMAP gateway state I think this does not belong here.

PepperJo commented on control/server.py:
```diff
@@ -127,10 +133,37 @@ def serve(self):
+                False: gateway_rpc.delete_listener
+            }
+        }
+        self.gateway_state.set_update_callbacks(callbacks)
```
I don't like that the `gateway_state` has to know about details of a RPC request. I think these callbacks should be more generic and they should contain all the logic of generating a RPC request. So there is no dependencies to protobufs etc. in `gateway_state`.

PepperJo commented on control/state.py:
```diff
@@ -9,95 +9,164 @@
+        self._remove_key(key)
 
     @abstractmethod
     def delete_state(self):
```
Is this used anywhere? Do we need it? If yes, we should make sure that there is a mechanism that someone doesn't accidentally remove the state.

sskaur replied:
I use it for testing purposes. We can move it to a test context in a different PR.

PepperJo reviewed (changes requested):
Thank you Sandy. My first set of comments.

trociny commented on control/state.py:
```diff
@@ -9,95 +9,164 @@
+        super().remove_subsystem(subsystem_nqn)
+
+        # Delete all keys related to subsystem
+        for key in self.state.keys():
```
It looks like if you replaced direct access to `self.state` with `self.get_state` you could avoid code duplication, defining `remove_subsystem` in the base class only.

trociny commented on control/state.py:
```diff
@@ -160,159 +266,227 @@ def _remove_key(self, key: str):
+               notify_id=None,
+               notifier_id=None,
+               watch_id=None,
+               data=None):
```
As all these unused args are actually the rados watcher specific, I would just define it as `def update(self)`, and then in `OmapGatewayState.update` I would register the watcher with a wrapper, something like below:
```
  def watcher_callback(notify_id,  notifier_id,  watch_id,  data):
    return callback()
  self.watch = self.ioctx.watch(self.omap_name, watcher_callback)
```

trociny commented on control/state.py:
```diff
@@ -160,159 +266,227 @@ def _remove_key(self, key: str):
+            # Update local state and version
+            self.local.update(remote_state_dict)
+            self.remote.set_local_version(int(remote_version))
+            self.logger.debug("Update complete.")
```
Wondering if the update would be so simple. E.g. I am not sure about how you handle `changed` keys. Suppose a bdev (rbd image) is changed. Actually, I suppose you can't just change image in bdev, you will need tor delete and create bdev, and for this you will need to recretead the namespace. So usually, when a "remote" gw is fast enough, it will just follow remove/add updates performed on the "local" gw. But suppose by some reason the "remote" gw missed some updates and now is catching up, so the bdev update is in `changed` now. Then it will try create_bdev, which will fail.

I might be not very clear, so let me know if you want me to elaborate. But my main thought is that I am not sure we may treat the `changed` the way you do it here. Probably we always need to delete/remove and then to create/add.

And the order of deleting may be different from the order of adding. E.g. we need to add a bdev before adding the namespace, but we can't remove the bdev before removing the namespace.

trociny reviewed (commented):
And it would ne  nice to have some multi-gateway tests.

PepperJo commented on control/grpc.py:
```diff
@@ -324,6 +324,7 @@ def remove_host(self, request, context=None):
     def create_listener(self, request, context=None):
         """Creates a listener for a subsystem at a given IP/Port."""
 
+        ret = False
```
Do we really want to return status `False` if we didn't create a listener on this particular node? I think we should succeed in that case. Listeners might be created elsewhere successfully with state update.

PepperJo commented on control/grpc.py:
```diff
@@ -380,6 +381,7 @@ def create_listener(self, request, context=None):
     def delete_listener(self, request, context=None):
         """Deletes a listener from a subsystem at a given IP/Port."""
 
+        ret = False
```
See above

PepperJo commented on control/__main__.py:
```diff
@@ -32,3 +32,4 @@
     config = GatewayConfig(args.config)
     with GatewayServer(config) as gateway:
         gateway.serve()
+        gateway.keep_alive()
```
Add new line after.

PepperJo commented on control/server.py:
```diff
@@ -91,14 +87,34 @@ def __exit__(self, exc_type, exc_value, traceback):
         # Register service implementation with server
+        omap_state = OmapGatewayState(self.config)
+        local_state = LocalGatewayState()
+        self.gateway_state = ConsistentGatewayState(self.config, local_state,
```
No need to make `gateway_state` a member instead use e.g. `GatewayState.BDEV_PREFIX` below

PepperJo commented on control/server.py:
```diff
@@ -243,3 +244,48 @@ def _ping(self):
+    def gateway_rpc_caller(self, requests, is_add_req):
+        """Passes RPC requests to gateway service."""
+        for key, val in requests.items():
+            if key.startswith(self.gateway_state.BDEV_PREFIX):
```
use `GatewayState.BDEV_PREFIX` (see above)

PepperJo commented on control/server.py:
```diff
@@ -243,3 +244,48 @@ def _ping(self):
+                                            pb2.delete_bdev_req(),
+                                            ignore_unknown_fields=True)
+                    self.gateway_rpc.delete_bdev(req)
+            elif key.startswith(self.gateway_state.SUBSYSTEM_PREFIX):
```
see above

PepperJo commented on control/server.py:
```diff
@@ -243,3 +244,48 @@ def _ping(self):
+                                            pb2.delete_subsystem_req(),
+                                            ignore_unknown_fields=True)
+                    self.gateway_rpc.delete_subsystem(req)
+            elif key.startswith(self.gateway_state.NAMESPACE_PREFIX):
```
see above

PepperJo commented on control/server.py:
```diff
@@ -243,3 +244,48 @@ def _ping(self):
+                                            pb2.remove_namespace_req(),
+                                            ignore_unknown_fields=True)
+                    self.gateway_rpc.remove_namespace(req)
+            elif key.startswith(self.gateway_state.HOST_PREFIX):
```
see above

PepperJo commented on control/server.py:
```diff
@@ -243,3 +244,48 @@ def _ping(self):
+                else:
+                    req = json_format.Parse(val, pb2.remove_host_req())
+                    self.gateway_rpc.remove_host(req)
+            elif key.startswith(self.gateway_state.LISTENER_PREFIX):
```
see above

PepperJo commented on control/state.py:
```diff
@@ -108,7 +180,7 @@ def __init__(self, config):
         conn = rados.Rados(conffile=ceph_conf)
         conn.connect()
         self.ioctx = conn.open_ioctx(ceph_pool)
```
I assume any of the 3 lines above can raise an exception so why not move it into the try block so we get some logging if it happens

PepperJo commented on control/state.py:
```diff
@@ -163,156 +272,214 @@ def _remove_key(self, key: str):
 
-    def _restore_bdevs(self, omap_dict, callback):
-        """Restores a bdev from the OMAP."""
+    def update(self, callback):
```
Why not start watch/notify via `__init__` or in a different function `register_watch` or similar ? `update` doesn't seem the right place to do this. The semantic for the other update state calls (other classes) is immediate but update here it is enabling some callback mechanism.

PepperJo commented on control/state.py:
```diff
@@ -163,156 +272,214 @@ def _remove_key(self, key: str):
-        self._add_key(key, val)
+    def add_subsystem(self, subsystem_nqn: str, val: str):
+        """Adds a subsystem to the state data store."""
+        self.remote.add_subsystem(subsystem_nqn, val)
```
If this fails because our local version is not up to date should we update and retry here or elsewhere? What if someone else added the same subsystem in the meantime. We already configured SPDK at this point so this is probably not going to work. Otherwise we would have to revert SPDK changes.

PepperJo replied:
Probably need to change the order to first update the state then update SPDK in case it fails revert the state change.

sskaur commented:
As the PR exists now, the multi-gateway functionality works when RPC requests are sent to a single gateway. An issue arises when requests are split between gateways, as a gateway will shut down with error when it receives a new request while in an outdated state. Any gateway will be outdated potentially up to the length of the version polling interval (state_update_interval_sec) every time another gateway in the group gets an incoming request. The implementation for update and retry instead of error for this case may require extended discussion. This could be a separate PR, and the current PR could get merged sooner with a statement about expected behavior if requests are split between gateways.

PepperJo commented on control/server.py:
```diff
@@ -91,14 +86,34 @@ def __exit__(self, exc_type, exc_value, traceback):
+        local_state = LocalGatewayState()
+        gateway_state = ConsistentGatewayState(self.config, local_state,
+                                               omap_state)
+        self.gateway_rpc = GatewayService(self.config, gateway_state,
```
Define in `__init__` and set to `None`

PepperJo commented on control/server.py:
```diff
@@ -91,14 +86,34 @@ def __exit__(self, exc_type, exc_value, traceback):
+        self._add_server_listener()
+
+        # Check for existing NVMeoF target state
+        gateway_state.start_update(self.gateway_rpc_caller)
```
start_update is only called on serve once so why not pass `gateway_rpc_caller` when creating the `ConsistentGatewayState` object? It seems that it only supports one gateway_rpc_caller anyway.

PepperJo commented on control/state.py:
```diff
@@ -7,109 +7,179 @@
 
-    @abstractmethod
-    def restore(self, callbacks):
+    def update(self):
```
Missing docstring

PepperJo commented on control/state.py:
```diff
@@ -7,109 +7,179 @@
+        """Deletes contents of local state object."""
+        self.state.clear()
+
+    def update(self, state_update):
```
Number of arguments differs from base

PepperJo commented on control/state.py:
```diff
@@ -125,9 +195,41 @@ def __init__(self, config):
+    def get_omap_version(self) -> str:
+        """Returns OMAP version."""
+        with rados.ReadOpCtx() as read_op:
+            iter, _ = self.ioctx.get_omap_vals_by_keys(read_op,
```
redefine builtin `iter` -> change name

PepperJo commented on control/state.py:
```diff
@@ -125,9 +195,41 @@ def __init__(self, config):
+                                                       (self.OMAP_VERSION_KEY,))
+            self.ioctx.operate_read_op(read_op, self.omap_name)
+            value_list = list(dict(iter).values())
+            if len(value_list) == 1:
```
move "parsing" outside `with`

PepperJo commented on control/state.py:
```diff
@@ -125,9 +195,41 @@ def __init__(self, config):
+    def get_state(self) -> Dict[str, str]:
+        """Returns dict of all OMAP keys and values."""
+        with rados.ReadOpCtx() as read_op:
+            iter, _ = self.ioctx.get_omap_vals(read_op, "", "", -1)
```
redefine builtin `iter` -> change name

PepperJo commented on control/state.py:
```diff
@@ -125,9 +195,41 @@ def __init__(self, config):
+            if len(value_list) == 1:
+                val = str(value_list[0], "utf-8")
+                return val
+        return None
```
Shouldn't we throw an exception if we encounter 0 or more than 1 omap version?

PepperJo commented on control/state.py:
```diff
@@ -125,9 +195,41 @@ def __init__(self, config):
+            self.ioctx.operate_read_op(read_op, self.omap_name)
+            value_list = list(dict(iter).values())
+            if len(value_list) == 1:
+                val = str(value_list[0], "utf-8")
```
Do we want to parse the version here directly as `int` instead since we seem to parse it later as `int` anyway?

PepperJo commented on control/state.py:
```diff
@@ -163,159 +270,214 @@ def _remove_key(self, key: str):
-        self._add_key(key, val)
+        if self.use_notify:  # Notify other gateways within the group of change
+            try:
+                notify = self.ioctx.notify(self.omap_name)
```
variable is never accessed

PepperJo commented on control/state.py:
```diff
@@ -144,9 +246,14 @@ def _add_key(self, key: str, val: str):
 
+        if self.use_notify:  # Notify other gateways within the group of change
+            try:
+                notify = self.ioctx.notify(self.omap_name)
```
variable is never accessed

PepperJo commented on control/state.py:
```diff
@@ -163,159 +270,214 @@ def _remove_key(self, key: str):
-        self._add_key(key, val)
+        try:
+            self.use_notify = True
+            self.watch = self.ioctx.watch(self.omap_name, _watcher_callback)
```
Define `watch` in `__init__` and set to `None`.
What if register_watch is called again? Should we check if `self.watch` is already set?

PepperJo commented on control/state.py:
```diff
@@ -163,159 +270,214 @@ def _remove_key(self, key: str):
+        self.local.delete_state()
+
+    def start_update(self, gateway_rpc_caller):
+        self.gateway_rpc_caller = gateway_rpc_caller
```
See comment where function is called

PepperJo commented on control/state.py:
```diff
@@ -163,159 +270,214 @@ def _remove_key(self, key: str):
+        self.remote.delete_state()
+        self.local.delete_state()
+
+    def start_update(self, gateway_rpc_caller):
```
Missing docstring

PepperJo commented on control/state.py:
```diff
@@ -163,159 +270,214 @@ def _remove_key(self, key: str):
+            self.remote.register_watch(self.update)
+
+        # Start polling for state updates
+        self.update_interval = self.config.getint("gateway",
```
Can we move this to `__init__` ?

sskaur replied:
I don't think so. We can't call for an update until the GatewayServer is instantiated, and we need an instance of ConsistentGatewayState to do this.

PepperJo replied:
I meant line 396 -> getting the update interval from the config. We don't expect it to change, no?

sskaur replied:
Ah, I see. Yes I can move that

PepperJo commented on control/state.py:
```diff
@@ -163,159 +270,214 @@ def _remove_key(self, key: str):
+        if self.update_interval < 1:
+            self.logger.info("Invalid state_update_interval_sec. Setting to 1.")
+            self.update_interval = 1
+        self.update_timer = threading.Thread(target=self._update_caller)
```
Define variable in `__init__` and set to `None`. Check if already set so in case function gets called again we don't get a 2nd thread.

PepperJo commented on control/state.py:
```diff
@@ -163,159 +270,214 @@ def _remove_key(self, key: str):
+        ]
+
+        remote_version = self.remote.get_omap_version()
+        if int(self.remote.get_local_version()) < int(remote_version):
```
See above. I think we can avoid converting to int here every time.

PepperJo commented on tests/conftest.py:
```diff
@@ -0,0 +1,24 @@
+
+
+@pytest.fixture(scope="session")
+def conffile(request):
```
docstring

PepperJo commented on tests/conftest.py:
```diff
@@ -0,0 +1,24 @@
+
+
+@pytest.fixture(scope="session")
+def config(conffile):
```
avoid using function name as parameter defined above. Docstring.

sskaur replied:
Using function names as parameters is done purposefully. These functions are fixtures, which provide setup and tear down for the test functions and [are called](https://docs.pytest.org/en/6.2.x/fixture.html) by identifying them in the parameter list.

The missing docstring was not done purposefully, I will add that.

PepperJo replied:
Ah got it. Thank you!

PepperJo commented on tests/conftest.py:
```diff
@@ -0,0 +1,24 @@
+
+
+@pytest.fixture(scope="session")
+def rbd_image(request):
```
docstring

PepperJo commented on tests/test_multi_gateway.py:
```diff
@@ -0,0 +1,99 @@
+
+
+@pytest.fixture(scope="module")
+def conn(config):
```
docstring

PepperJo commented on tests/test_multi_gateway.py:
```diff
@@ -0,0 +1,99 @@
+    gatewayB.gateway_state.delete_state()
+
+
+def test_multi_gateway_coordination(config, rbd_image, conn):
```
docstring. Avoid `conn` parameter because of same name for function above

PepperJo commented on tests/test_multi_gateway.py:
```diff
@@ -0,0 +1,99 @@
+    ret_bdev = stubA.create_bdev(bdev_req)
+    ret_subsystem = stubA.create_subsystem(subsystem_req)
+    ret_namespace = stubA.add_namespace(namespace_req)
+    assert ret_bdev.status == True
```
use `is True` (also below)

PepperJo commented on tests/test_state.py:
```diff
@@ -0,0 +1,162 @@
+    ioctx.close()
+
+
+def init_state(config, update_interval_sec, notify):
```
docstring

PepperJo commented on tests/test_state.py:
```diff
@@ -0,0 +1,162 @@
+
+
+@pytest.fixture(scope="module")
+def ioctx(config):
```
avoid `ioctx` function name that is used as parameter below

PepperJo reviewed (changes requested):
Some minor comments. Otherwise LGTM

trociny commented on control/server.py:
```diff
@@ -224,16 +225,15 @@ def _create_transport(self, trtype):
-        }
-        self.gateway_state.restore(callbacks)
+    def keep_alive(self):
+        """Continuously confirm confirm communication with SPDK process."""
```
nit: extra "confirm". And if we want to follow the style of other doc comments, it should be "confirms".

trociny commented on control/state.py:
```diff
@@ -163,159 +270,214 @@ def _remove_key(self, key: str):
+            # Handle OMAP additions
+            added_keys = remote_state_dict.keys() - local_state_dict.keys()
+            added = {key: remote_state_dict[key] for key in added_keys}
+            sorted_added = self._update_sort(added, prefix_list)
```
nit: names like `grouped_added` and `_group_by_prefix`  would be more intuitive for me (may be due to `SQL` background).

trociny commented on control/state.py:
```diff
@@ -163,159 +270,214 @@ def _remove_key(self, key: str):
+                        nsid = key.rsplit("_", 1)[1]
+                        val_dict = json.loads(val)
+                        val_dict['nsid'] = nsid
+                        val = json.dumps(val_dict)
```
nit: Just wonder, don't we store `nsid` in the `val` already? If we don't then why not? We would avoid this exercise here.

trociny commented on control/state.py:
```diff
@@ -163,159 +270,214 @@ def _remove_key(self, key: str):
+            # Update local state and version
+            self.local.update(remote_state_dict)
+            self.remote.set_local_version(int(remote_version))
+            self.logger.debug("Update complete.")
```
nit: Just wonder, if there was some special reason you decided to add new states first and then remove the old ones? Because I would expect that removing first and then adding would work better. And  then I would just add the `changed` keys to both `added` and `removed` lists.

I still have feeling it might not work as expected in some age cases, but I may be wrong and I think it is good enough anyway and we can improve later if needed.

trociny reviewed (commented):
I added some comments, though nothing important that could block the merge.

majianpeng commented:
@sskaur how to understand gateway group? In a group, all gateway(w/ spdk) are the same, like raid1 or ? Is there any doc for gateway group. Thanks!

PepperJo commented:
> @sskaur how to understand gateway group? In a group, all gateway(w/ spdk) are the same, like raid1 or ? Is there any doc for gateway group. Thanks!

Hi @majianpeng. You can check out this talk which explains the grouping concept: https://www.youtube.com/watch?v=vrqVi2Bdfp4

sskaur commented:
@PepperJo @trociny I believe I've addressed all your comments in my recent changes. Please take another look and let me know if you have more feedback.

PepperJo commented on control/state.py:
```diff
@@ -125,9 +195,45 @@ def __init__(self, config):
 
+    def __exit__(self, exc_type, exc_value, traceback):
+        if self.use_notify:
+            self.watch.close()
```
if `register_watch` was never called this will fail. Instead of checking on `use_notify` check if `watch` is `None`.

PepperJo reviewed (changes requested):
Only one minor comment otherwise LGTM

trociny reviewed (approved):
Seems good to me. I have not tried it yet and could miss something during review, still I think the sooner we merge it the better not to slow down development. It will not hurt anybody if there is a bug and we will be able to fix it quickly.

Reviewed-by: Mykola Golub <mykola.golub@clyso.com>

PepperJo reviewed (approved):
Thank you. LGTM

idryomov commented on control/grpc.py:
```diff
@@ -145,8 +143,7 @@ def create_subsystem(self, request, context=None):
 
-        self.logger.info(
-            f"Received request to delete {request.subsystem_nqn}")
+        self.logger.info(f"Received request to delete {request.subsystem_nqn}")
```
These changes appear to be unrelated -- it's generally better to separate pure formatting stuff.  But, since it's being touched anyway, "delete" -> "delete subsystem":
```suggestion
        self.logger.info(
            f"Received request to delete subsystem {request.subsystem_nqn}")
```

sskaur replied:
In the future, I'll separate out formatting.

idryomov commented on control/grpc.py:
```diff
@@ -196,6 +193,11 @@ def add_namespace(self, request, context=None):
+                    # Update request with NSID
+                    dict_req = json.loads(json_req)
+                    dict_req['nsid'] = str(nsid)
+                    json_req = json.dumps(dict_req)
```
Instead of decoding and then immediately encoding almost the same JSON, why not test whether `request.nsid` is set before calling `MessageToJson`?  It seems like `request.nsid` can be set right there, if needed.

sskaur replied:
Sounds good.

idryomov commented on control/state.py:
```diff
@@ -163,159 +274,218 @@ def _remove_key(self, key: str):
-        self._add_key(key, val)
+        if self.watch is None:
+            try:
+                self.use_notify = True
```
This assignment appears to be redundant as `register_watch` call is itself conditioned on `use_notify`.

sskaur replied:
👍

idryomov commented on control/state.py:
```diff
@@ -7,109 +7,179 @@
+
+    def _add_key(self, key: str, val: str):
+        """Adds key and value to the local state object."""
+        self.state[key] = val.encode('utf-8')
```
What is the purpose of this `encode` call?  If it's to match what `ioctx.set_omap()` does internally then it's probably wrong because `ioctx.get_omap_vals` also does the reverse (i.e. decodes) internally.  Contrast that with how this dictionary is being returned directly from `LocalGatewayState.get_state` a few lines above.

sskaur replied:
I'll remove this encode.

idryomov commented on control/state.py:
```diff
@@ -7,109 +7,179 @@
+    LISTENER_PREFIX = "listener_"
+
+    def get_state(self) -> Dict[str, str]:
+        """Returns state specification."""
```
"specification" is really vague and can mean a ton of different things:
```suggestion
        """Returns the state dictionary."""
```

sskaur replied:
👍

idryomov commented on control/state.py:
```diff
@@ -7,109 +7,179 @@
-        pass
+
+class LocalGatewayState(GatewayState):
+    """Persists gateway NVMeoF target state to a local object.
```
I wouldn't use the word "persists" when it comes to in-memory state.  Also, "local object" is pretty confusing:
```suggestion
    """Records gateway NVMeoF target state in a dictionary.
```

sskaur replied:
👍

idryomov commented on control/state.py:
```diff
@@ -7,109 +7,179 @@
+    """Persists gateway NVMeoF target state to a local object.
+
+    Records persistent NVMeoF target state data in key/value format within a 
+    local object.
```
Ditto here and other docstrings in this class.

sskaur replied:
👍

idryomov commented on ceph-nvmeof.conf:
```diff
@@ -14,6 +14,8 @@ group =
 addr = 127.0.0.1
 port = 5500
 enable_auth = False
+state_update_notify = False
```
Why is the use of watch/notify disabled by default?  Isn't it supposed to be the workhorse, with periodic updates being there just as a fallback?

sskaur replied:
I'll change to enable by default.

idryomov commented on control/state.py:
```diff
@@ -163,159 +274,218 @@ def _remove_key(self, key: str):
+
+        remote_version = self.remote.get_omap_version()
+        if self.remote.get_local_version() < remote_version:
+            local_state_dict = self.local.get_state()
```
Similarly, is there protection against `local_state_dict` getting updated while we are scanning for changes below?  In particular, the list of keys is constructed three times.

idryomov replied:
The list of keys is now constructed once (good!) but I think the question still stands because `get_state` here returns a reference to a dictionary which can be mucked with while `update` is running.  For example, if a key is removed immediately after `local_state_keys = local_state_dict.keys()`, wouldn't the gateway crash on an exception thrown by `local_state_dict[key]` in "Find OMAP changes"?

sskaur replied:
You're right, I missed that.

sskaur replied:
@idryomov I changed `get_state` to return a shallow copy.

idryomov commented on control/state.py:
```diff
@@ -163,159 +274,218 @@ def _remove_key(self, key: str):
+    def __init__(self, config, local, remote, gateway_rpc_caller):
+        self.config = config
+        self.local = local
+        self.remote = remote
```
I'd suggest renaming `remote` to `omap`.  I don't know if it's just me, but referring to `OmapGatewayState` as `remote` seems misleading because when you are looking at a piece of code like
```
        self.remote.add_namespace(subsystem_nqn, nsid, val)
        self.local.add_namespace(subsystem_nqn, nsid, val)
```
it's very easy to interpret it as this gateway _directly_ updating state of some other gateway, which is not how it actually works.
```suggestion
        self.omap = omap
```

sskaur replied:
👍

idryomov commented on control/state.py:
```diff
@@ -7,109 +7,179 @@
+        """Deletes contents of local state object."""
+        self.state.clear()
+
+    def update(self, state_update):
```
Let's avoid confusion with `ConsistentGatewayState.update` and make the signature reflect how this method is actually used:
```suggestion
    def reset(self, omap_state):
```

sskaur replied:
👍

idryomov commented on control/state.py:
```diff
@@ -163,159 +274,218 @@ def _remove_key(self, key: str):
+
+        if self.use_notify:
+            # Register a watch on remote state
+            self.remote.register_watch(self.update)
```
It looks like there can be two updates running at the same time in different threads: one induced by a notification and one by a periodic timer.  I think some serialization is in order here.

One way to do it would be to make the notification callback just wake the periodic timer thread and ensure that `update_time` timestamp is adjusted accordingly (i.e. make notification-induced updates count towards `update_interval` as well).

sskaur replied:
👍

idryomov commented on control/server.py:
```diff
@@ -243,3 +246,48 @@ def _ping(self):
+                else:
+                    req = json_format.Parse(val,
+                                            pb2.delete_bdev_req(),
+                                            ignore_unknown_fields=True)
```
Can you explain the addition of `ignore_unknown_fields` -- why it's specified for `delete_bdev_req`, `delete_subsystem_req` and `remove_namespace_req` but not for `remove_host_req` or `delete_listener_req`.

sskaur replied:
The original `create_X_req` / `add_X_req` is used to generate the respective delete/remove request when an update identifies its removal from the OMAP. Since the create/add requests for bdev, subsystem, and namespace have more fields than their delete/remove counterparts, those can be ignored when building the delete requests.

idryomov commented on control/state.py:
```diff
@@ -7,109 +7,179 @@
+        """Returns state specification."""
         pass
 
-    @abstractmethod
```
Shouldn't `get_state`, `_add_key` and `_remove_key` be marked as abstract?

sskaur replied:
Only `delete_state` is implemented in all the `GatewayState` subclasses.

idryomov replied:
This actually highlights the weirdness of this inheritance hierarchy.  `LocalGatewayState` and `OmapGatewayState` are the only real implementations of `GatewayState` and they both implement `get_state`, `_add_key` and `_remove_key`.  Any class that meaningfully implements `GatewayState` _has_ to implement all of them as they are what defines the "state".

`ConsistentGatewayState` just orchestrates updates of `LocalGatewayState` and `OmapGatewayState` and shouldn't inherit from `GatewayState` in the first place: the fact that `get_state` can't be defined for `ConsistentGatewayState` because there is no actual "state" speaks to this IMO.  And I would actually rename it to something like `GatewayStateHandler` to avoid the confusion.

sskaur replied:
Sure, makes sense.

idryomov commented on control/state.py:
```diff
@@ -7,109 +7,179 @@
 #  Authors: anita.shekar@ibm.com, sandy.kaur@ibm.com
 #
 
+import json
```
Not seeing any JSON processing in this file.

sskaur replied:
👍

idryomov commented on control/state.py:
```diff
@@ -163,159 +274,218 @@ def _remove_key(self, key: str):
+            self.HOST_PREFIX, self.LISTENER_PREFIX
+        ]
+
+        remote_version = self.remote.get_omap_version()
```
Not sure if this is part of what is being punted to the future as per https://github.com/ceph/ceph-nvmeof/pull/29#issuecomment-1355249582, but I think the version and the rest of the config should be fetched in one go.  Otherwise an update can be missed, or, worse, some updated state can be applied and tagged with the wrong version.

sskaur replied:
👍

idryomov reviewed (commented):
Please update the description of the PR and the commit message with the list of limitations (no racing requests, etc).  I see https://github.com/ceph/ceph-nvmeof/pull/29#issuecomment-1355249582 but it needs to be made more visible.

idryomov commented on control/state.py:
```diff
@@ -144,9 +244,14 @@ def _add_key(self, key: str, val: str):
             self.logger.error(f"Unable to add key to omap: {ex}. Exiting!")
             raise
 
+        if self.watch:  # Notify other gateways within the group of change
```
I think the gateway should notify unconditionally, even if `state_update_notify` is disabled in its config.  Otherwise, `state_update_notify = False` in one gateway's config would affect all gateways in the gateway group.

idryomov commented on control/state.py:
```diff
@@ -163,159 +268,221 @@ def _remove_key(self, key: str):
-        """Adds a bdev to the OMAP."""
-        key = self.BDEV_PREFIX + bdev_name
-        self._add_key(key, val)
+        if self.watch:  # Notify other gateways within the group of change
```
ditto

idryomov commented on control/state.py:
```diff
@@ -163,159 +268,221 @@ def _remove_key(self, key: str):
+            update_time = time.time() + self.update_interval
+            self.update()
+            notify_event.wait(max(update_time - time.time(), 0))
+            if notify_event.is_set():
```
Is calling `is_set` necessary here?

sskaur replied:
I'll remove it.

sskaur committed:
initial multi-gateway support

This commit allows for coordination of the target state within all
gateways in a gateway group. This functionality is only supported when
all requests are sent to a single gateway. Sending requests to mutliple
gateways within the group may result in error if a gateway receives a
new request while in an outdated state. With the state_update_notify
config option, gateways register a watch on the group's OMAP object and
send notify signals to the object after state changes. Gateways check
for OMAP updates upon receiving the watch callback as well as
periodically according to an state_update_interval_sec config parameter.

Signed-off-by: Sandy Kaur <sandy.kaur@ibm.com>

sskaur commented:
@idryomov @PepperJo @trociny Let me know if you have further change requests. I believe this PR is ready for a merge.

idryomov commented:
> I believe this PR is ready for a merge.

Please rebase to get rid of the merge commit.

sskaur commented:
> > I believe this PR is ready for a merge.
> 
> Please rebase to get rid of the merge commit.

Done, thanks.

idryomov commented on tests/test_state.py:
```diff
@@ -0,0 +1,171 @@
+                                _state_polling_update)
+    state.update_interval = update_interval_sec
+    state.use_notify = False
+    key = "bdev_test"
```
I don't think `GatewayStateHandler` change had to lead to `bdev` being hard-coded here.  I'm disregarding this for now since it's test code, but I think we need to consolidate all on-disk format constants under a single entity and use them consistently in production and test.  It doesn't make sense to separate `OMAP_VERSION_KEY` (currently under `OmapGatewayState`) and e.g. `BDEV_PREFIX` (currently under `GatewayState`) when both are used in/define the same RADOS object.

Please send a follow up PR for that.

epuertat commented:
@sskaur I'm not fully sure it's related to this PR, but I just rebased #90 and it previously kept running:
```
ceph-nvmeof-devel    | DEBUG:control.server:create_transport: tcp options: 
ceph-nvmeof-devel    | [2023-04-05 12:18:05.880409] tcp.c: 629:nvmf_tcp_create: *NOTICE*: *** TCP Transport Init ***
ceph-nvmeof-devel    | INFO:control.state:First gateway: created object nvmeof.state
ceph-nvmeof-devel    | DEBUG:control.state:Read key: omap_version -> 1
ceph-nvmeof-devel    | INFO:control.state:This omap was just created. Nothing to restore
ceph-nvmeof-devel    | INFO:control.server:Starting server...
```

But it now automatically finishes...: 

```
ceph-nvmeof          | DEBUG:control.server:create_transport: tcp options: 
ceph-nvmeof          | [2023-04-05 17:06:15.301635] tcp.c: 629:nvmf_tcp_create: *NOTICE*: *** TCP Transport Init ***
ceph-nvmeof          | INFO:control.state:First gateway: created object nvmeof.state
ceph-nvmeof          | INFO:control.server:Terminating SPDK...
ceph-nvmeof          | INFO:control.server:Exiting the gateway process.
```

Do you have any idea about what's going on there? You can easily reproduce by checking out my PR and just running `docker-compose up ceph-nvmeof-devel`?
